src/network.rs

```diff
@@ -0,0 +1,508 @@
+	// Create a public/private key pair, either based on a seed or random
+	let id_keys = match seed {
+		Some(seed) => {
+			let mut bytes = [0u8; 32];
+			bytes[0] = seed;
+			let secret_key = ed25519::SecretKey::from_bytes(&mut bytes)
+				.expect("Error should only appear if length is wrong.");
+			identity::Keypair::Ed25519(secret_key.into())
+		},
+		None => identity::Keypair::generate_ed25519(),
+	};
+	let local_peer_id = id_keys.public().to_peer_id();
+	info!("Local peer id: {:?}", local_peer_id);
+
+	// try to get psk
+	let psk: Option<PreSharedKey> = get_psk(psk_path)?
+		.map(|text| PreSharedKey::from_str(&text))
+		.transpose()?;
+	// create transport
+	let transport = setup_transport(id_keys, psk);
+
+	// create swarm that manages peers and events
+	let swarm = {
+		let mut kad_cfg = KademliaConfig::default();
+		kad_cfg.set_query_timeout(Duration::from_secs(5 * 60));
+		let store_cfg = MemoryStoreConfig {
+			max_records: 24000000, // ~2hrs
+			max_value_bytes: 100,
+			max_providers_per_key: 1,
+			max_provided_keys: 100000,
+		};
+		let kad_store = MemoryStore::with_config(local_peer_id, store_cfg);
+		let behaviour = NetworkBehaviour {
+			kademlia: Kademlia::with_config(local_peer_id, kad_store, kad_cfg),
+		};
+
+		// Build the Swarm, connecting the lower transport logic with the
+		// higher layer network behaviour logic
+		SwarmBuilder::new(transport, behaviour, local_peer_id)
+			// connection background tasks are spawned onto the tokio runtime
+			.executor(Box::new(|fut| {
+				tokio::spawn(fut);
+			}))
+			.build()
+	};
+
+	let (command_sender, command_receiver) = mpsc::channel(10000);
+	let (event_sender, event_receiver) = mpsc::channel(10000);
+
+	Ok((
+		Client {
+			sender: command_sender,
+		},
+		ReceiverStream::new(event_receiver),
+		EventLoop::new(swarm, command_receiver, event_sender),
+	))
+}
+
+/// Read the pre-shared key file from the given directory
+fn get_psk(location: &String) -> Result<Option<String>> {
+	let path = Path::new(location);
+	let swarm_key_file = path.join("swarm.key");
+	match fs::read_to_string(swarm_key_file) {
+		Ok(text) => Ok(Some(text)),
+		Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(None),
+		Err(e) => Err(e.into()),
+	}
+}
+
+fn setup_transport(
+	key_pair: Keypair,
+	psk: Option<PreSharedKey>,
+) -> transport::Boxed<(PeerId, StreamMuxerBox)> {
+	let noise = NoiseAuthenticated::xx(&key_pair).unwrap();
+
+	let base_transport =
+		TokioTcpTransport::new(GenTcpConfig::default().nodelay(true).port_reuse(false));
+	let maybe_encrypted = match psk {
+		Some(psk) => EitherTransport::Left(
+			base_transport.and_then(move |socket, _| PnetConfig::new(psk).handshake(socket)),
+		),
+		None => EitherTransport::Right(base_transport),
+	};
+
+	maybe_encrypted
+		.upgrade(Version::V1)
+		.authenticate(noise)
+		.multiplex(SelectUpgrade::new(
+			YamuxConfig::default(),
+			MplexConfig::new(),
+		))
+		.timeout(Duration::from_secs(20))
```

Comment:
What's the role of this timeout? Should we parametarize its value?

Reply:
This is a timeout to the setup and protocol upgrade process for all inbound and outbound connections established through the transport.
I think this is nothing critical ATM, but we could add it to the config later, if needed.